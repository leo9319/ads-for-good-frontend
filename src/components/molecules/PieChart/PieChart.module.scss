@import '../../../styles/colours';
@import '../../../styles/typography';

.chartWrapper {
  &.size1 .chartCircleWrapper {
    width: 320px;
    height: 320px;
  }

  &.size2 .chartCircleWrapper {
    width: 160px;
    height: 160px;
  }

  &.size3 .chartCircleWrapper {
    width: 80px;
    height: 80px;
  }
}

.chartCircleWrapper {
  position: relative;
  display: flex;
  align-items: center;
  justify-content: center;
  margin-bottom: var(--spacing-xs);
}

.chartLabel {
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  text-align: center;
  padding: 0 var(--spacing-sm);
  max-width: 85%;
  display: flex;
  flex-direction: column;
  align-items: center;

  div {
    margin-bottom: var(--spacing-2xs);
    color: $color-neutral-900;
    font-weight: var(--typography-font-weight-bold);
  }

  p {
    margin: 0;
    color: $color-neutral-700;
    word-wrap: break-word;
    overflow: hidden;
    text-overflow: ellipsis;
    text-align: center;

    display: -webkit-box;
    -webkit-line-clamp: 3;
    -webkit-box-orient: vertical;

    display: -moz-box;
    -moz-line-clamp: 3;
    -moz-box-orient: vertical;

    display: box;
    box-orient: vertical;
    line-clamp: 3;
  }

  &.size1 {
    div {
      @extend .heading-lg;
    }

    p {
      @extend .body-md;
    }
  }

  &.size2 {
    div {
      @extend .heading-sm;
    }

    p {
      @extend .body-xs;
    }
  }

  &.size3 {
    div {
      @extend .heading-xs;
    }

    p {
      @extend .body-xs;
    }
  }
}
