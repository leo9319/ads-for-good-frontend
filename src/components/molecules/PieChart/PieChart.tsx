import React from 'react';
import styles from './PieChart.module.scss';
import { classNames } from '@utils/common/classNames';

export interface DonationImpactProps {
  value: number; // percentage
  description: string; // replaces title
  color: string;
  size?: '1' | '2' | '3'; // "1" = 80px, "2" = 160px, "3" = 320px
}

export interface PieChartProps {
  data: DonationImpactProps;
  className?: string;
}

export const PieChart = ({ data, className }: PieChartProps) => {
  const { value, description, color, size = '1' } = data;

  const radius = 50;
  const strokeWidth = 25;
  const circumference = 2 * Math.PI * radius;
  const offset = circumference * (1 - value / 100);

  // Reverse size mapping: "1" = size3 (80px), "2" = size2 (160px), "3" = size1 (320px)
  const sizeClassMap = {
    '1': 'size3',
    '2': 'size2',
    '3': 'size1',
  };
  const appliedSize = sizeClassMap[size];

  const wrapperClass = classNames(
    styles.chartWrapper,
    styles[appliedSize],
    className
  );
  const labelClass = classNames(styles.chartLabel, styles[appliedSize]);

  const truncatedDescription =
    description.length > 40
      ? description.slice(0, 40).trim() + 'â€¦'
      : description;

  return (
    <div className={wrapperClass}>
      <div className={styles.chartCircleWrapper}>
        <svg width="100%" height="100%" viewBox="0 0 140 140">
          <circle
            cx="70"
            cy="70"
            r={radius}
            stroke="#eee"
            strokeWidth={strokeWidth}
            fill="none"
          />
          <circle
            cx="70"
            cy="70"
            r={radius}
            stroke={color}
            strokeWidth={strokeWidth}
            fill="none"
            strokeDasharray={circumference}
            strokeDashoffset={offset}
            strokeLinecap="butt"
            transform="rotate(-90 70 70)"
          />
        </svg>

        <div className={labelClass}>
          <div>
            <strong>{value}%</strong>
          </div>
          <p>{truncatedDescription}</p>
        </div>
      </div>
    </div>
  );
};

export default PieChart;
