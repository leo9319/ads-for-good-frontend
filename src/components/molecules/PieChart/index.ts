export { default } from './PieChart';
export { PieChart } from './PieChart';
export type { PieChartProps, DonationImpactProps } from './PieChart';
