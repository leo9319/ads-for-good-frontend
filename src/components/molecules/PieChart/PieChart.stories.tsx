import type { Meta, StoryObj } from '@storybook/react';
import PieChart from './PieChart';

const meta: Meta<typeof PieChart> = {
  title: 'Library/Molecules/PieChart',
  component: PieChart,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {
    data: {
      control: 'object',
      description:
        'Donation impact entry with value (percentage), description, color, and size (1=Small, 2=Medium, 3=Large)',
    },
    className: {
      control: { type: 'text' },
      description: 'Optional class name to style the container',
    },
  },
};

export default meta;
type Story = StoryObj<typeof meta>;

export const Large: Story = {
  args: {
    data: {
      value: 84.9,
      description: 'Field programs and advocacy',
      color: '#ec6707',
      size: '3', // Large
    },
    className: '',
  },
};

export const Medium: Story = {
  args: {
    data: {
      value: 60,
      description: 'Emergency response',
      color: '#ff6b6b',
      size: '2', // Medium
    },
    className: '',
  },
};

export const Small: Story = {
  args: {
    data: {
      value: 42,
      description: '',
      color: '#4db8ff',
      size: '1', // Small
    },
    className: '',
  },
};
