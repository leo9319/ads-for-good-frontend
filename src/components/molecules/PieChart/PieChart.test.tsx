import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom'; // ✅ Required for custom matchers like toBeInTheDocument
import PieChart, { DonationImpactProps } from './PieChart';

const mockData: DonationImpactProps = {
  value: 75,
  description:
    'This is a test description to check the truncation logic of the text inside the circle',
  color: '#ff6347',
  size: '2',
};

describe('Chart component', () => {
  it('renders the chart with the correct percentage value', () => {
    render(<PieChart data={mockData} />);
    expect(screen.getByText(/75%/)).toBeInTheDocument();
  });

  it('applies the correct size class based on the "size" prop', () => {
    const { container } = render(<PieChart data={mockData} />);
    expect(container.firstChild).toHaveClass('size2');
  });

  it('truncates description to 40 characters followed by ellipsis', () => {
    render(<PieChart data={mockData} />);
    const truncatedText = screen.getByText(
      /This is a test description to check the/
    );
    expect(truncatedText).toHaveTextContent(
      'This is a test description to check the…'
    );
  });

  it('renders the background circle and progress arc correctly', () => {
    const { container } = render(<PieChart data={mockData} />);
    const circles = container.querySelectorAll('circle');
    expect(circles.length).toBe(2);
    expect(circles[0]).toHaveAttribute('stroke', '#eee');
    expect(circles[1]).toHaveAttribute('stroke', mockData.color);
  });

  it('matches snapshot for size2', () => {
    const { container } = render(<PieChart data={mockData} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
