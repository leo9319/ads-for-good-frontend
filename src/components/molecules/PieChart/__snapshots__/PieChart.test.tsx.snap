// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chart component matches snapshot for size2 1`] = `
<div
  class="chartWrapper size2"
>
  <div
    class="chartCircleWrapper"
  >
    <svg
      height="100%"
      viewBox="0 0 140 140"
      width="100%"
    >
      <circle
        cx="70"
        cy="70"
        fill="none"
        r="50"
        stroke="#eee"
        stroke-width="25"
      />
      <circle
        cx="70"
        cy="70"
        fill="none"
        r="50"
        stroke="#ff6347"
        stroke-dasharray="314.1592653589793"
        stroke-dashoffset="78.53981633974483"
        stroke-linecap="butt"
        stroke-width="25"
        transform="rotate(-90 70 70)"
      />
    </svg>
    <div
      class="chartLabel size2"
    >
      <div>
        <strong>
          75
          %
        </strong>
      </div>
      <p>
        This is a test description to check theâ€¦
      </p>
    </div>
  </div>
</div>
`;

exports[`PieChart Default 1`] = `
<div
  class="chartWrapper size1"
>
  <div
    class="chartCircleWrapper"
  >
    <svg
      height="100%"
      viewBox="0 0 140 140"
      width="100%"
    >
      <circle
        cx="70"
        cy="70"
        fill="none"
        r="50"
        stroke="#eee"
        stroke-width="25"
      />
      <circle
        cx="70"
        cy="70"
        fill="none"
        r="50"
        stroke="#ec6707"
        stroke-dasharray="314.1592653589793"
        stroke-dashoffset="47.344022806988865"
        stroke-linecap="butt"
        stroke-width="25"
        transform="rotate(-90 70 70)"
      />
    </svg>
    <div
      class="chartLabel size1"
    >
      <div>
        <strong>
          84.9%
        </strong>
      </div>
      <p>
        Field programs and advocacy
      </p>
    </div>
  </div>
</div>
`;
